### PR TITLE
Coming soon: Do not render header and footer if the theme does not support FSE

### LINKED ISSUE #50915: wp_head seems to be called twice on coming soon pages
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

I just had the issue that on the coming soon page of a category archive i got an error of a custom function called twice.

I don't encounter the issue on any other page.

I am not using a full site editing template.

### Expected behavior

wp_head should only be executed once.

### Actual behavior

It seems that wp_head is executed another time when coming_soon is enabled in the settings.

### Steps to reproduce

1. Create a fresh site with WooCommerce 9.2.3
2. Installed and activated the `Twenty Twenty-One` theme
3. Added the following code via a custom plugin:

```php
function test() { echo 'test-head'; }
add_action('wp_head','test');
```

4. Enabled the coming soon page from Site visibility settings
5. Turn on `Restrict to store pages only` option
6. Go to the front-end /shop page
7. Confirm that the `test-head` is displayed twice

Screenshot 2024-09-20 at 18 11 48

### WordPress Environment

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.2.2
Legacy REST API Package Version: Auf dieser Website ist das alte REST-API-Plugin nicht installiert.
Action Scheduler Version: ✔ 3.8.1
Log Directory Writable: ✔
WP Version: 6.6.1
WP Multisite: –
WP Memory Limit: 1 GB
WP Debug Mode: ✔
WP Cron: ✔
Language: de_DE
External object cache: –

### Server Environment ###

Server Info: nginx/1.26.1
PHP Version: 8.2.22
PHP Post Max Size: 100 MB
PHP Time Limit: 600
PHP Max Input Vars: 5000
cURL Version: 7.88.1
OpenSSL/3.0.13

SUHOSIN Installed: –
MySQL Version: 8.0.36-0ubuntu0.22.04.1
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

acf: 1
acf-field: 97
acf-field-group: 13
attachment: 1157
events: 14
faq: 52
gnomeposts: 8
highlight: 3
nav_menu_item: 44
oembed_cache: 1
page: 64
popup: 4
popup_theme: 8
post: 2
product: 5
recipebanner: 8
recipes: 655
revision: 1523
stories: 7
team: 26
wp_global_styles: 1
wp_navigation: 1
wpsl_stores: 2948

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Fehlermeldungen sollten Besuchern nicht angezeigt werden.

### Active Plugins (19) ###

ACF Content Analysis for Yoast SEO: von Thomas Kräftner
ViktorFroberg
marol87
pekz0r
angrycreative
Team Yoast – 3.0.1

Admin and Site Enhancements (ASE): von wpase.com – 7.3.0
Advanced Custom Fields PRO: von WP Engine – 6.3.4
Autoptimize: von Frank Goossens (futtta) – 3.1.12
bs Swiper: von Bootscore – 5.8.3
Image Placeholders: von WordPress Performance Team – 1.1.1
EWWW Image Optimizer: von Exactly WWW – 7.8.0
Performance Lab: von WordPress Performance Team – 3.4.0
Post Types Order: von Nsp Code – 2.2.4
Regenerate Thumbnails: von Alex Mills (Viper007Bond) – 3.1.6
Rating Recipes: von risingBYTES GmbH –
Trusted Shops Easy Integration for WooCommerce: von vendidero – 2.0.3
Modern Image Formats: von WordPress Performance Team – 2.1.0
German Market: von MarketPress – 3.39
WooCommerce: von Automattic – 9.2.2
Yoast SEO: von Team Yoast – 23.2
WP Migrate: von WP Engine – 2.6.10
WP Store Locator - CSV Manager: von Tijmen Smit – 1.2.12
WP Store Locator: von Tijmen Smit – 2.2.241

### Inactive Plugins (15) ###

Category and Taxonomy Meta Fields: von Aftab Husain – 1.0.0
DG Optimize: von Dynamo Gold – 0.1
Geolocation IP Detection: von Yellow Tree (Benjamin Pick) – 4.2.3
Popup Maker: von Popup Maker – 1.16.3
Popup Maker - Advanced Targeting Conditions: von WP Popup Maker – 1.4.5
Popup Maker - Advanced Theme Builder: von WP Popup Maker – 1.1.6
Popup Maker - Exit Intent Popups: von WP Popup Maker – 1.3.1
Popup Maker - MailChimp Integration: von WP Popup Maker – 1.3.3
Popup Maker - Scroll Triggered Popups: von WP Popup Maker – 1.3.2
ProfilePress: von ProfilePress Team – 3.2.5
Quick Page/Post Redirect Plugin: von anadnet – 5.2.4
Soil: von Roots – 4.0.4
W3 Total Cache: von BoldGrid – 2.2.1
WP Import Export Lite: von VJInfotech – 3.9.15
WP Performance Tester: von <a href="https://reviewsignal.com">Kevin Ohashi</a> – 2.0.0

### Dropin Plugins () ###

object-cache.php: Performance Lab Server Timing Object Cache Drop-In

### Must Use Plugins (13) ###

DG BIGO Capabilities: von dynamo gold – 0.1
DG BIGO CPT Event: von dynamo gold – 0.1
DG BIGO CPT FAQ: von dynamo gold – 0.1
DG BIGO CPT Gnome: von dynamo gold – 0.1
DG BIGO CPT Gnome: von dynamo gold – 0.1
DG BIGO CPT Highlights: von dynamo gold – 0.1
DG BIGO CPT Recipes: von dynamo gold – 0.1
DG BIGO CPT Success Story: von dynamo gold – 0.1
DG BIGO CPT Team: von dynamo gold – 0.1
DG BIGO Image Sizes: von dynamo gold – 0.1
DG BIGO Tax Recipes: von dynamo gold – 0.1
DG BIGO WPSL Modifications: von dynamo gold – 0.1
WP Migrate Compatibility: von Delicious Brains – 1.3

### Settings ###

Legacy API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: left_space
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 Tage
Level threshold: –
Log directory size: 51 kB

### WC Pages ###

Shop-Basis: #8505 - /shop/
Warenkorb: #8506 - /cart/ -  Enthält den woocommerce/cart-Block
Kasse: #8507 - /checkout/ -  Enthält den woocommerce/checkout-Block
Mein Konto: #8508 - /my-account/
Allgemeine Geschäftsbedingungen: #8539 - /allgemeine_geschaeftsbedingungen/

### Theme ###

Name: Twenty Twenty-Two
Version: 1.8
Author URL: https://de.wordpress.org/
Child Theme: ❌ – Wenn du WooCommerce mit einem Theme
das du nicht selbst entwickelt hast
an eigene Bedürfnisse anpasst
dann empfehlen wir die Verwendung eines Child-Themes. Siehe Wie du ein Child-Theme erstellst (engl.)

WooCommerce Support: ✔

### Templates ###

Overrides: /var/www/html/www/wp-content/plugins/woocommerce/templates/block-notices/error.php
/var/www/html/www/wp-content/plugins/woocommerce/templates/block-notices/notice.php
/var/www/html/www/wp-content/plugins/woocommerce/templates/block-notices/success.php


### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: experimental-blocks
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system

Daily Cron: ✔ Next scheduled: 2024-08-24 09:47:49 +02:00
Options: ✔
Notes: 67
Onboarding: completed

### Action Scheduler ###

Abgeschlossen: 20
Oldest: 2024-08-11 11:49:10 +0200
Newest: 2024-08-23 13:26:12 +0200

Fehlgeschlagen: 1
Oldest: 2024-08-19 16:41:14 +0200
Newest: 2024-08-19 16:41:14 +0200

Ausstehend: 1
Oldest: 2024-08-24 10:42:29 +0200
Newest: 2024-08-24 10:42:29 +0200


### Status report information ###

Generated at: 2024-08-23 13:44:07 +02:00
```
</details>

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by moon0326**
…current theme is not a fse

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #50915  .

This PR fixes double rendering of the header and footer sections when the current theme does not support FSE and the `store pages only` option is enabled for the coming soon mode.

### Cause

When the current theme does not support FSE and the `store pages only` option is enabled, we defer the rendering the header and footer to the theme by calling [get_header](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/ComingSoon/ComingSoonRequestHandler.php#L70) and [get_footer](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/ComingSoon/ComingSoonRequestHandler.php#L85). However, we also render the `coming-soon` content in the [template-canvas.php](https://github.com/WordPress/WordPress/blob/master/wp-includes/template-canvas.php), which includes both the header and footer.

Note: Calling [get_query_tempalte( 'coming-soon' )](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/ComingSoon/ComingSoonRequestHandler.php#L67) returns template-canvas.php. This is expected.

Screenshot 2024-09-23 at 2 01 43 PM

We can fix it by rendering the coming soon template content without the header and footer part when the theme is not FSE and the store pages only option is enabled.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new JN site with this branch.
2. Install and activate Code Snippets plugin.
3. Go to `Snippets -> Add New` and create a new snippet with the following code. Make sure to enable the snippet.

```
function test() { echo 'test-head'; }
add_action('wp_head','test');
```
4. Install and activate Twenty Twenty-One theme (non-FSE theme).
5. Go to `WooCommerce -> Settings -> Site Visibility`
6. Enable `Coming soon` with `Apply to store pages only` option.
7. Open a new incognito window and navigate to https://YOUR-SITE/shop
8. You should see only one `test-head` text.
 

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Fix double rendering of the header and footer when the coming soon mode is enabled with the store pages only option with a non-FSE theme.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/ComingSoon/ComingSoonRequestHandler.php [hunk 1]
```diff
@@ -78,7 +78,11 @@ function () {
 		);
 
 		if ( ! empty( $coming_soon_template ) && file_exists( $coming_soon_template ) ) {
-			include $coming_soon_template;
+			if ( ! $is_fse_theme && $is_store_coming_soon ) {
+				echo get_the_block_template_html();
```
Comment (moon0326) [inline plugins/woocommerce/src/Internal/ComingSoon/ComingSoonRequestHandler.php:83]: This function is marked as private on https://developer.wordpress.org/reference/functions/get_the_block_template_html/, but it is the only function that I could find to render the template content without custom code.

I believe it is relatively safe to use, as it was introduced back in WordPress 5.8.0.

Comment (chihsuan) [reply plugins/woocommerce/src/Internal/ComingSoon/ComingSoonRequestHandler.php:83]: Nice find! I think this is a good solution. 🙌  How about we add a check to see if the function exists before calling it? This way, we can avoid any potential issues if the function is removed in the future.

**Review by chihsuan — APPROVED** [review]
Tested well! 👍  Just left a small comment about the function_exists check.

**Review by psealock — APPROVED** [review]
This is testing well, approved pending @chihsuan's comment

**Comment by moon0326** [comment]
> How about we add a check to see if the function exists before calling it? This way, we can avoid any potential issues if the function is removed in the future.

The function is defined in [block-templates.php](https://github.com/WordPress/WordPress/blob/master/wp-includes/block-template.php#L218) and it is included on every page load in [wp-settings.php](https://github.com/WordPress/WordPress/blob/master/wp-settings.php#L197C24-L197C49) without any conditions. I think we can safely use it.

> the function is removed in the future.

Update:  I missed this part. Working on it!

@chihsuan  Updated in [Make sure get_the_block_template_html exist](https://github.com/woocommerce/woocommerce/pull/51633/commits/d3b404205bea8cb91f28b973ad72b698513b408a)

**Review by chihsuan — APPROVED** [review]
LGTM, thanks!

**Pull request merged**